# Data extraction

In [ ]:
# Install pymed: a Python library that provides access to PubMed through the PubMed API.
# https://github.com/gijswobben/pymed

pip install pymed

In [42]:
from pymed import PubMed
import pandas as pd
import numpy as np
import csv

In [2]:
# Query form generator for pubmed: https://pubmed.ncbi.nlm.nih.gov/advanced/

pubmed = PubMed()

my_query = """
("The Journal of neuroscience : the official journal of the Society for Neuroscience"[Journal])
AND ("2011"[Date - Publication] : "2020"[Date - Publication])
"""

results = pubmed.query(my_query, max_results=15000)

In [3]:
print(results)

- 'results' is an intertools.chain object of size > 10,000 rows

In [ ]:
# Write a new CSV file using 'csv.writer'
# First step: convert the text into a dictionary
# Second step: write a new row of .csv file by using 'writerow' method
# Result: at every iter, a new row will be written in the csv
# Why do this: it's faster than appending each row to a list or a dictionary...

def create_csv(iterable, filename): # filename: output csv filename
    filename_csv = (filename+'.'+'csv')
    
    with open(filename_csv, 'w') as csvfile:
        writer = csv.writer(csvfile)

        for it in iterable:
            it_dict = it.toDict()
            writer.writerow(val for key, val in it_dict.items())


create_csv(results, 'jneuro')

In [4]:
# Extract column names

for result in results:
    result_dict = result.toDict()
    if len(result_dict) > 0:
        break

columns = [key for key, val in result_dict.items()]
print(columns)

['pubmed_id', 'title', 'abstract', 'keywords', 'journal', 'publication_date', 'authors', 'methods', 'conclusions', 'results', 'copyrights', 'doi', 'xml']


In [5]:
# Import the csv as a dataframe

jneuro = pd.read_csv('../data/jneuro.csv', names=columns)

print(jneuro.shape)
jneuro.head()

(12734, 13)


pubmed_id                                              title  \
0  32737169  Erratum: Salido and Ramamurthy, "Proteoglycan ...   
1  32737168  Erratum: Keitel et al., "Medial Nucleus Accumb...   
2  32737167  The neocortical progenitor specification progr...   
3  32732324  Flexible coordinator and switcher hubs for ada...   
4  32727820  Responses to Visual Speech in Human Posterior ...   

                                            abstract keywords  \
0                                                NaN       []   
1                                                NaN       []   
2  Neuronal progenitors in the developing forebra...       []   
3  Functional connectivity studies have identifie...       []   
4  Experimentalists studying multisensory integra...       []   

                                             journal publication_date  \
0  The Journal of neuroscience : the official jou...       2020-08-02   
1  The Journal of neuroscience : the official jou...       2020-08-02   
2  The Journal of neuroscience : the official jou...       2020-08-02   
3  The Journal of neuroscience : the official jou...       2020-08-01   
4  The Journal of neuroscience : the official jou...       2020-07-31   

                                             authors  methods  conclusions  \
0                                                 []      NaN          NaN   
1                                                 []      NaN          NaN   
2  [{'lastname': 'Yabut', 'firstname': 'Odessa R'...      NaN          NaN   
3  [{'lastname': 'Cocuzza', 'firstname': 'Carrisa...      NaN          NaN   
4  [{'lastname': 'Metzger', 'firstname': 'Brian A...      NaN          NaN   

  results                       copyrights                             doi  \
0     NaN                              NaN  10.1523/JNEUROSCI.1846-20.2020   
1     NaN                              NaN  10.1523/JNEUROSCI.1829-20.2020   
2     NaN    Copyright © 2020 the authors.  10.1523/JNEUROSCI.2888-19.2020   
3     NaN    Copyright © 2020 the authors.  10.1523/JNEUROSCI.2559-19.2020   
4     NaN  Copyright © 2020 Metzger et al.  10.1523/JNEUROSCI.0279-20.2020   

                                           xml  
0  <Element 'PubmedArticle' at 0x7f18e0bb0a10>  
1  <Element 'PubmedArticle' at 0x7f18e07a1050>  
2  <Element 'PubmedArticle' at 0x7f18e07d5b90>  
3  <Element 'PubmedArticle' at 0x7f18e07c2a10>  
4  <Element 'PubmedArticle' at 0x7f18e07cc710>

## Observation:
- 12734 articles from 2011-01-07 to 2020-08-02
- The data is sorted by 'publication_date (YYYY-MM-DD)' in a descending order by default

###   